core/sys/linux/00_linux.odin

```diff
@@ -0,0 +1,144 @@
+// most parameters are going to be trivially-castable to
+// uptr, so we'll have that.
+@private
+syscall0 :: #force_inline proc "contextless" (nr: uptr)->iptr {
+	return cast(iptr) intrinsics.syscall(nr)
+}
+
+@private
+syscall1 :: #force_inline proc "contextless" (nr: uptr, p1: $T)->iptr
+where
+	size_of(p1) <= size_of(uptr)
+{
+	return cast(iptr) intrinsics.syscall(nr, cast(uptr) p1)
+}
+
+@private
+syscall2 :: #force_inline proc "contextless" (nr: uptr, p1: $T1, p2: $T2)->iptr
+where
+	size_of(p1) <= size_of(uptr) &&
+	size_of(p2) <= size_of(uptr) 
+{
+	return cast(iptr) intrinsics.syscall(nr, cast(uptr) p1, cast(uptr) p2)
+}
+
+@private
+syscall3 :: #force_inline proc "contextless" (nr: uptr, p1: $T1, p2: $T2, p3: $T3)->iptr
+where
+	size_of(p1) <= size_of(uptr) &&
+	size_of(p2) <= size_of(uptr) &&
+	size_of(p3) <= size_of(uptr)
+{
+	return cast(iptr) intrinsics.syscall(nr, cast(uptr) p1, cast(uptr) p2, cast(uptr) p3)
+}
+
+@private
+syscall4 :: #force_inline proc "contextless" (nr: uptr, p1: $T1, p2: $T2, p3: $T3, p4: $T4)->iptr
+where
+	size_of(p1) <= size_of(uptr) &&
+	size_of(p2) <= size_of(uptr) &&
+	size_of(p3) <= size_of(uptr) &&
+	size_of(p4) <= size_of(uptr)
+{
+	return cast(iptr) intrinsics.syscall(nr, cast(uptr) p1, cast(uptr) p2, cast(uptr) p3, cast(uptr) p4)
+}
+
+@private
+syscall5 :: #force_inline proc "contextless" (nr: uptr, p1: $T1, p2: $T2, p3: $T3, p4: $T4, p5: $T5)->iptr
+where
+	size_of(p1) <= size_of(uptr) &&
+	size_of(p2) <= size_of(uptr) &&
+	size_of(p3) <= size_of(uptr) &&
+	size_of(p4) <= size_of(uptr) &&
+	size_of(p5) <= size_of(uptr)
+{
+	return cast(iptr) intrinsics.syscall(nr, cast(uptr) p1, cast(uptr) p2, cast(uptr) p3, cast(uptr) p4, cast(uptr) p5)
+}
+
+@private
+syscall6 :: #force_inline proc "contextless" (nr: uptr, p1: $T1, p2: $T2, p3: $T3, p4: $T4, p5: $T5, p6: $T6)->iptr
+where
+	size_of(p1) <= size_of(uptr) &&
+	size_of(p2) <= size_of(uptr) &&
+	size_of(p3) <= size_of(uptr) &&
+	size_of(p4) <= size_of(uptr) &&
+	size_of(p5) <= size_of(uptr) &&
+	size_of(p6) <= size_of(uptr)
+{
+	return cast(iptr) intrinsics.syscall(nr, cast(uptr) p1, cast(uptr) p2, cast(uptr) p3, cast(uptr) p4, cast(uptr) p5, cast(uptr) p6)
+}
+
+syscall :: proc {syscall0, syscall1, syscall2, syscall3, syscall4, syscall5, syscall6}
+
+// Note(flysand): 32-bit architectures sometimes take in a 64-bit argument in a
+// register pair. This function should help me avoid typing the same code a few times..
+when LONG_BITS == 32 {
+	// xxx64 system calls take some parameters as pairs of ulongs rather than a single pointer
+	@private compat64_arg_pair :: #force_inline proc "contextless" (a: i64)->(hi: c_ulong, lo: c_ulong) {
+		no_sign := uint(a)
+		hi = c_ulong(no_sign >> 32)
+		lo = c_ulong(no_sign & 0xffff_ffff)
+		return
+	}
+} else {
+	// ... and on 64-bit architectures it's just a long
+	@private compat64_arg_pair :: #force_inline proc "contextless" (a: i64)->(c_ulong) {
+		return c_ulong(a)
+	}
+}
```

Comment:
Where is this used, and why do they have different signatures?

Why not this?
```odin
@private compat64_arg_pair :: #force_inline proc "contextless" (a: i64) -> (hi: c_ulong, lo: c_ulong) {
	when LONG_BITS == 32 {
		no_sign := uint(a)
		hi = c_ulong(no_sign >> 32)
		lo = c_ulong(no_sign & 0xffff_ffff)
	} else {
		lo = c_ulong(a)
		hi = 0
	}
	return
}
```

Reply:
Different return types. We don't want to use 7th register when calling `syscall` (I don't remember on the top of my head where that might cause the issue, actually probably won't it's just a waste of 4 bytes on the stack), nor I want to clobber an extra register on a syscall when it gets inlined.

It's not performance-critical but my mind wouldn't feel at ease if I pass extra parameter for no reason. It's the same reason why I'm doing polymorphic procedure groups for syscalls where a parameter is ignored depending on one of the arguments (+ the type safety)